llvm/include/llvm/TableGen/StringToOffsetTable.h

```diff
@@ -52,6 +52,33 @@ class StringToOffsetTable {
     return II->second;
   }
 
+  // Emit the string using string literal concatenation, for better readability
+  // and searchability.
+  void EmitStringLiteralDef(raw_ostream &OS, const Twine &Decl,
+                            const Twine &Indent = "  ") const {
+    OS << "\n"
+       << Indent << "#ifdef __GNUC__\n"
+       << Indent << "#pragma GCC diagnostic push\n"
+       << Indent << "#pragma GCC diagnostic ignored \"-Woverlength-strings\"\n"
+       << Indent << "#endif\n"
+       << Indent << Decl << " = {\n";
```

Comment:
The declaration and braces should be printed outside so that this function could be used in other contexts.

Reply:
Ah, and the braces are not really needed.

Reply:
The pragma's are needed to avoid clang warnings. Most tables emitted will needed them in general, which is why they are in the emitter. May be we should have a function to emit these pragmas somewhere and reuse them here and in `SequenceToOffsetTable`? What's a good place for such a function?

Reply:
How about: static class functions in StringToTableOffset (which is in llvm/TableGen/StringToOffsetTable.h) to emit the pragmas, make emit string just emit the data (and call it `EmitStringAsFragments` or something), and reuse the static functions in `SequenceToOffsetTable` (in a follow up)?

Reply:
It does violate [rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) so maybe just keep it as it is?

Reply:
So these 3 functions:

* `StringToTableOffset::EmitStringAsFragments(raw_ostream &OS, StringRef Indent)`
* `StringToTableOffset::PushIgnoreStringLength(raw_ostream &OS, StringRef Indent)`
* `StringToTableOffset::PopIgnoreStringLength(raw_ostream &OS, StringRef Indent)`

Reply:
As is in emit pragmas in EmitStringAsFragments()? or as separate function and call from IntrinsicEmitter? I think one of your comments was "declaration and braces should be printed outside so that this function could be used in other contexts". I guess you mean inline the pragma emission in IntrinsicEmitter?

Reply:
> I guess you mean inline the pragma emission in IntrinsicEmitter?

What I meant is that the call site should look like:

```
    OS << "  static constexpr char BuiltinNames[] = \n";
    Table.EmitStringAsFragments(OS);
    OS << ";\n\n";
```

I didn't consider pragmas.

I now see that SequenceToOffsetTable works similar to your implementation, so I think it is fine to keep them aligned.

Reply:
I could rename the function to EmitAsStringFragments(), may be its more descriptive than EmitStringLiteralDef?

Reply:
Nah, its fine. It does emit a definition, so the name isn't wrong.